crates/engine/tree/src/tree/precompile_cache.rs

```diff
@@ -0,0 +1,188 @@
+    mini_moka::sync::CacheBuilder::new(100_000)
+        .build_with_hasher(alloy_primitives::map::DefaultHashBuilder::default())
+}
+
+/// Cache key, just input for each precompile.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct CacheKey(alloy_primitives::Bytes);
+
+/// Combined entry containing both the result and gas bounds.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CacheEntry {
+    /// The actual result of executing the precompile.
+    result: PrecompileResult,
+    /// Observed gas used on successful run
+    gas_used: u64,
+    /// Gas limit observed on call error.
+    err_gas_limit: u64,
+}
+
+/// A cache for precompile inputs / outputs.
+///
+/// This assumes that the precompile is a standard precompile, as in `StandardPrecompileFn`, meaning
+/// its inputs are only `(Bytes, u64)`.
+///
+/// NOTE: This does not work with "context stateful precompiles", ie `ContextStatefulPrecompile` or
+/// `ContextStatefulPrecompileMut`. They are explicitly banned.
+#[derive(Debug)]
+pub(crate) struct CachedPrecompile {
+    /// Cache for precompile results and gas bounds.
+    cache: Arc<PrecompileCache>,
+    /// The precompile.
+    precompile: DynPrecompile,
+    /// Cache metrics.
+    metrics: CachedPrecompileMetrics,
+}
+
+impl CachedPrecompile {
+    /// `CachedPrecompile` constructor.
+    pub(crate) fn new(precompile: DynPrecompile, cache: Arc<PrecompileCache>) -> Self {
+        Self { precompile, cache, metrics: Default::default() }
+    }
+
+    pub(crate) fn wrap(precompile: DynPrecompile, cache: Arc<PrecompileCache>) -> DynPrecompile {
+        let wrapped = Self::new(precompile, cache);
+        move |data: &Bytes, gas_limit: u64| -> PrecompileResult { wrapped.call(data, gas_limit) }
+            .into()
+    }
+
+    fn increment_by_one_precompile_cache_hits(&self) {
+        self.metrics.precompile_cache_hits.increment(1);
+    }
+
+    fn increment_by_one_precompile_cache_misses(&self) {
+        self.metrics.precompile_cache_misses.increment(1);
+    }
+
+    fn cache_entry_count(&self) -> u64 {
+        self.cache.entry_count()
+    }
+
+    fn update_precompile_cache_size(&self, previous_entry_count: u64) {
+        let new_entry_count = self.cache.entry_count();
+        if new_entry_count > previous_entry_count {
+            self.metrics
+                .precompile_cache_size
+                .increment((new_entry_count - previous_entry_count) as f64);
+        }
+    }
+}
+
+impl Precompile for CachedPrecompile {
+    fn call(&self, data: &Bytes, gas_limit: u64) -> PrecompileResult {
+        let key = CacheKey(data.clone());
+
+        let cache_result = self.cache.get(&key);
+
+        if let Some(ref entry) = cache_result {
+            match &entry.result {
+                Ok(_) => {
+                    self.increment_by_one_precompile_cache_hits();
+                    if gas_limit < entry.gas_used {
+                        return Err(PrecompileError::OutOfGas)
+                    }
```

Comment:
I'd be more comfortable here if we invoke the actual precompile for this (rare) case

Reply:
makes sense, we only cache successful calls now and return the output if the gas limit is bigger than the previous gas used, otherwise the precompile is always executed